### PR TITLE
Allow add-hoc metrics and template (and Add first version of standalone example of dataset with LLM as a judge )

### PR DESCRIPTION
**Pull request by eladven**
… metric

### CODE REVIEW

File: examples/standalone_evaluation_llm_as_judge.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+)
+
+add_to_catalog(
+    rating_template,
+    "templates.rating_template",
+    overwrite=True,
+)
+
+rating_metric = LLMAsJudge(
+    inference_model=HFPipelineBasedInferenceEngine(
```
Comment (yoavkatz) [inline examples/standalone_evaluation_llm_as_judge.py:40]: I would use a  strong large mode in service  IBMGenAI (or OpenAI) in the example.

File: examples/standalone_evaluation_llm_as_judge.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        model_name="google/flan-t5-large", max_new_tokens=256, use_fp16=True
+    ),
+    template="templates.rating_template",
+    task="rating.single_turn",
+    format="formats.models.mistral.instruction",
+    main_score="flan-t5-large_huggingface",
+    strip_system_prompt_and_format_from_inputs=False,
```
Comment (yoavkatz) [inline examples/standalone_evaluation_llm_as_judge.py:47]: Why is this required?

File: examples/standalone_evaluation_llm_as_judge.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+
+add_to_catalog(
+    rating_metric,
+    "metrics.rating_metric",
+    overwrite=True,
+)
+
+card = TaskCard(
+    loader=LoadFromDictionary(data=data),
+    task=Task(
```
Comment (yoavkatz) [inline examples/standalone_evaluation_llm_as_judge.py:58]: I would move in this example to use the catalog (https://github.com/IBM/unitxt/blob/03ef303b280fdc0453b23db8620a1f0482a9a8b0/examples/qa_evaluation.py#L21) and override the metic.

This will make it shorter, and also demonstrate how one overrides the metric.

File: examples/standalone_evaluation_llm_as_judge.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    ' by strictly following this format: "[[rating]]"'
+    ".\n\n",
+    input_format="[Question]\n{question}\n\n" "[Assistant's Answer]\n{answer}\n",
+    output_format="[[{rating}]]",
+    postprocessors=[
+        r"processors.extract_mt_bench_rating_judgment",
+    ],
+)
+
+add_to_catalog(
```
Comment (yoavkatz) [inline examples/standalone_evaluation_llm_as_judge.py:33]: Best if we can avoid storing in the catalog - and just pass as object.  Note it requires a fix as in #920.

File: src/unitxt/artifact.py [hunk 5]
```diff
@@ -404,13 +407,27 @@ def __str__(self):
...
+def fetch_artifact(artifact_rep):
+    if isinstance(artifact_rep, Artifact):
+        return artifact_rep, None
+    if Artifact.is_artifact_file(artifact_rep):
+        return Artifact.load(artifact_rep), None
 
-    artifactory, name, args = get_artifactory_name_and_args(name=name)
-
-    return artifactory.get_with_overwrite(name, overwrite_args=args), artifactory
+    try:
```
Comment (yoavkatz) [inline src/unitxt/artifact.py:416]: Instead of try catch, use the method that checks legal catalog artifact name.

File: src/unitxt/artifact.py [hunk 6]
```diff
@@ -279,9 +279,12 @@ def __post_init__(self):
     def _to_raw_dict(self):
         return {"type": self.type, **self._init_dict}
 
-    def save(self, path):
+    def to_json(self):
         data = self.to_dict()
-        save_json(path, data)
+        return json_dump(data)
+
```
Comment (yoavkatz) [inline src/unitxt/artifact.py:285]: def serialize(self):
    if ( __id___) return __id___ else self.to_json()

File: src/unitxt/artifact.py (hunk 5 above)
Comment (yoavkatz) [reply src/unitxt/artifact.py:416]: Take 413-430 become Artifact.deserialize(str)

**Comment by yoavkatz** [comment]
Elad - when completed also update unitxt/tests/library/test_api.py evaluate tests, which currently does not check serialization of post processors, because it's not done correctly.

File: examples/standalone_evaluation_llm_as_judge.py (hunk 2 above)
Comment (yoavkatz) [reply examples/standalone_evaluation_llm_as_judge.py:54]: Why should it be false.  Doesn't it work with 'true'?

File: examples/standalone_evaluation_llm_as_judge.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+    input_format="[Question]\n{question}\n\n" "[Assistant's Answer]\n{answer}\n",
+    output_format="[[{rating}]]",
+    postprocessors=[
+        r"processors.extract_mt_bench_rating_judgment",
+    ],
+)
+
+platform = "hf"
+model_name = "google/flan-t5-large"
+inference_model = HFPipelineBasedInferenceEngine(
```
Comment (elronbandel) [inline examples/standalone_evaluation_llm_as_judge.py:37]:
```suggestion
inference_model = HFPipelineInferenceEngine(
```

File: examples/standalone_evaluation_llm_as_judge.py (hunk 1 above)
Comment (eladven) [reply examples/standalone_evaluation_llm_as_judge.py]: I added Lamma (in commented out section

File: examples/standalone_evaluation_llm_as_judge.py (hunk 2 above)
Comment (eladven) [reply examples/standalone_evaluation_llm_as_judge.py:54]: As we discussed, currently the code supports only rating.single_turn (and rating.single_turn_with_reference), The main score name has been changed. 
Regarding strip_system_prompt_and_format_from_inputs. Currently setting it to true caused the run to crash. I think we should solve it, but in a separate PR.

File: src/unitxt/artifact.py (hunk 5 above)
Comment (eladven) [reply src/unitxt/artifact.py]: Done

File: examples/standalone_evaluation_llm_as_judge.py (hunk 3 above)
Comment (yoavkatz) [reply examples/standalone_evaluation_llm_as_judge.py:59]: It's a tradeoff - we do want people to use catalog tasks - and we already have an example like this.  We don't want to repeat this on every example and we want to show new capabilities .  Still,   If you feel strong about it though, I'm fine.

File: examples/standalone_evaluation_llm_as_judge.py (hunk 2 above)
Comment (yoavkatz) [reply examples/standalone_evaluation_llm_as_judge.py:54]: Ok.

File: examples/standalone_evaluation_llm_as_judge.py (hunk 7 above)
Comment (yoavkatz) [reply examples/standalone_evaluation_llm_as_judge.py:37]: Best  be in a different PR.

File: examples/standalone_evaluation_llm_as_judge.py [hunk 8]
```diff
@@ -0,0 +1,104 @@
...
+                input_format="{question}",
+                output_format="{answer}",
+                postprocessors=["processors.lower_case"],
+            )
+        }
+    ),
+)
+
+# Convert card to a dataset
+recipe = StandardRecipe(card=card, template_card_index="simple")
```
Comment (yoavkatz) [inline examples/standalone_evaluation_llm_as_judge.py:78]: These 3 lines can now be done with load_dataset(card=card, template_card_index="simple").

File: examples/standalone_evaluation_llm_as_judge.py [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+
+# Infer a model to get predictions.
+model_name = "google/flan-t5-base"
+inference_model = HFPipelineBasedInferenceEngine(
+    model_name=model_name, max_new_tokens=32
+)
+predictions = inference_model.infer(test_dataset)
+
+# Evaluate the predictions using the defined metric.
+dataset_with_scores = evaluate(predictions=predictions, data=test_dataset)
```
Comment (yoavkatz) [inline examples/standalone_evaluation_llm_as_judge.py:91]: Now that predictions and processed predictions are returns.  These lines can be set done via:

```
evaluated_dataset = evaluate(predictions=predictions, data=test_dataset)

# Print results
for instance in evaluated_dataset:
    logger.info("*" * 80)
    logger.info(f"Model input:\n{instance['source']}")
    logger.info(
        f"Model prediction (as returned by the model):\n{instance['prediction']}"
    )
    logger.info(
        f"Model prediction (after post processing):\n{instance['processed_prediction']}"
    )
    logger.info(f"References:\n{instance['references']}")
    logger.info(
        f"References (after post processing):\n{instance['processed_references']}"
    )
    score_name = instance["score"]["instance"]["score_name"]
    score = instance["score"]["instance"]["score"]
    logger.info(f"Sample score ({score_name}) : {score}")
global_score = evaluated_dataset[0]["score"]["global"]["score"]
logger.info("*" * 80)
logger.info(f"Aggregated score ({score_name}) : {global_score}")
```

This is simpler.

File: src/unitxt/dataset_utils.py [hunk 10]
```diff
@@ -20,13 +21,18 @@ def parse(query: str):
...
     register_all_artifacts()
-    recipe = fetch(dataset_str)
+    recipe = fetch(dataset)
     if recipe is None:
-        args = parse(dataset_str)
-        if "__type__" not in args:
-            args["__type__"] = settings.default_recipe
+        args = parse(dataset)
+        if "type" not in args:
+            args["type"] = settings.default_recipe
```
Comment (elronbandel) [inline src/unitxt/dataset_utils.py:36]: Has changed on main to `__type__`

File: src/unitxt/llm_as_judge.py [hunk 11]
```diff
@@ -121,22 +122,22 @@ def compute(
...
-            recipe = f"{recipe},system_prompt={self.system_prompt}"
+            recipe_args["system_prompt"] = self.system_prompt
         if self.format:
-            recipe = f"{recipe},format={self.format}"
-
+            recipe_args["format"] = self.format
+        recipe = Artifact.from_dict(recipe_args)
         dataset = produce(instances, recipe)
         verdicts = self.inference_model.infer(dataset)
         meta_scores = evaluate(predictions=verdicts, data=dataset)
```
Comment (elronbandel) [inline src/unitxt/llm_as_judge.py:139]: Why is this necessary? Just for the postprocessors utilisation?

**Review by elronbandel — COMMENTED** [review]
See my comments. In general I think we should have post_process function like produce that can do so without calculating metrics (for no reason). Also it would be useful for others to have a function like infer(instances, recipe, inference_engine) that can bring all this utility in one line (this was requested by a team using paraphrasing for augmentations and would love to utilise all our tools for that end to end.

File: examples/standalone_evaluation_llm_as_judge.py (hunk 8 above)
Comment (eladven) [reply examples/standalone_evaluation_llm_as_judge.py:78]: Done. Indeed much clearer

File: examples/standalone_evaluation_llm_as_judge.py (hunk 9 above)
Comment (eladven) [reply examples/standalone_evaluation_llm_as_judge.py:91]: Done. It is indeed simpler, but please note that the evaluated_dataset doesn't contain 'processed_prediction' field. Therefore we avoid printing the prediction before the processing (I am not sure, but I think that the prediction field is after the postprocessing).

File: src/unitxt/dataset_utils.py (hunk 10 above)
Comment (eladven) [reply src/unitxt/dataset_utils.py:36]: Done. Thanks for letting me know.

File: src/unitxt/llm_as_judge.py (hunk 11 above)
Comment (eladven) [reply src/unitxt/llm_as_judge.py:139]: The code here is using unitxt to build a dataset for the judge inference. The old code used a representation of string, that represents dict, that represents a recipe.  In addition to the fact that this way is much less clear and bug-prone, it doesn't allow to use ad-hoc template.

Comment (elronbandel) [reply src/unitxt/llm_as_judge.py:139]: Agree, the question is about the use of `evaluate`, is it just for the post-processing aspect?

Comment (yoavkatz) [reply src/unitxt/llm_as_judge.py:139]: I believe so.  Right now, there is no simple way to run the post processors without executing the metrics.

File: examples/standalone_evaluation_llm_as_judge.py [hunk 12]
```diff
@@ -0,0 +1,105 @@
...
+model_name = "google/flan-t5-base"
+inference_model = HFPipelineBasedInferenceEngine(
+    model_name=model_name, max_new_tokens=32
+)
+predictions = inference_model.infer(test_dataset)
+
+# Evaluate the predictions using the defined metric.
+evaluated_dataset = evaluate(predictions=predictions, data=test_dataset)
+
+for instance in evaluated_dataset:
```
Comment (yoavkatz) [inline examples/standalone_evaluation_llm_as_judge.py:90]: Simplify with new printouts format using print_dict()

**Review by yoavkatz — APPROVED** [review]

**Review by yoavkatz — CHANGES_REQUESTED** [review]
Also - don't forget to update the example documentation.

File: examples/standalone_evaluation_llm_as_judge.py (hunk 12 above)
Comment (eladven) [reply examples/standalone_evaluation_llm_as_judge.py:90]: done

**Review by yoavkatz — APPROVED** [review]
